File: app/src/main/java/eu/darken/bluemusic/main/core/service/VolumeObserver.java
package eu.darken.bluemusic.main.core.service;

import android.database.ContentObserver;
import android.os.Handler;
import android.util.SparseArray;
import android.util.SparseIntArray;

import javax.inject.Inject;
import javax.inject.Named;

import timber.log.Timber;


public class VolumeObserver extends ContentObserver {

    private final StreamHelper streamHelper;

    interface Callback {
        void onVolumeChanged(int streamId, int volume);
    }

    private final SparseArray<Callback> callbacks = new SparseArray<>();
    private final SparseIntArray volumes = new SparseIntArray();

    @Inject
    public VolumeObserver(@Named("VolumeObserver") Handler handler, StreamHelper streamHelper) {
        super(handler);
        this.streamHelper = streamHelper;
    }

    public void addCallback(int streamType, Callback callback) {
        callbacks.put(streamType, callback);
        final int volume = streamHelper.getCurrentVolume(streamType);
        volumes.put(streamType, volume);
    }

    @Override
    public boolean deliverSelfNotifications() {
        return super.deliverSelfNotifications();
    }

    @Override
    public void onChange(boolean selfChange) {
        super.onChange(selfChange);
        Timber.v("Change detected.");
        for (int i = 0; i < callbacks.size(); i++) {
            int streamType = callbacks.keyAt(i);
            Callback callback = callbacks.get(streamType);
            int newVolume = streamHelper.getCurrentVolume(streamType);
            int oldVolume = volumes.get(streamType);
            if (newVolume != oldVolume) {
                Timber.v("Volume changed (type=%d, old=%d, new=%d)", streamType, oldVolume, newVolume);
                volumes.put(streamType, newVolume);
                callback.onVolumeChanged(streamType, newVolume);
            }
        }
    }
}

File: app/src/main/java/eu/darken/bluemusic/main/core/audio/VolumeObserver.kt
package eu.darken.bluemusic.main.core.audio

import android.database.ContentObserver
import android.os.Handler
import android.os.HandlerThread
import eu.darken.bluemusic.common.debug.logging.log
import eu.darken.bluemusic.common.debug.logging.logTag
import eu.darken.bluemusic.common.debug.logging.Logging.Priority.*
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

@Singleton
class VolumeObserver @Inject constructor(
    private val streamHelper: StreamHelper
) : ContentObserver(
    run {
        val handlerThread = HandlerThread("VolumeObserver")
        handlerThread.start()
        val looper = handlerThread.looper
        Handler(looper)
    }
) {

    interface Callback {
        fun onVolumeChanged(streamId: AudioStream.Id, volume: Int)
    }

    private val callbacks = mutableMapOf<AudioStream.Id, Callback>()
    private val volumes = mutableMapOf<AudioStream.Id, Int>()

    fun addCallback(id: AudioStream.Id, callback: Callback) {
        callbacks[id] = callback
        val volume = streamHelper.getCurrentVolume(id)
        volumes[id] = volume
    }

    override fun deliverSelfNotifications(): Boolean {
        return false
    }

    override fun onChange(selfChange: Boolean) {
        super.onChange(selfChange)
        log(TAG, VERBOSE) { "Change detected." }
        for ((id, callback) in callbacks) {
            val newVolume = streamHelper.getCurrentVolume(id)
            val oldVolume = volumes[id] ?: -1
            if (newVolume != oldVolume) {
                log(TAG, VERBOSE) { "Volume changed (type=$id, old=$oldVolume, new=$newVolume)" }
                volumes[id] = newVolume
                callback.onVolumeChanged(id, newVolume)
            }
        }
    }

    companion object {
        private val TAG = logTag("VolumeObserver")
    }
}